### PR TITLE
Updated the interface documentation

### PR DESCRIPTION
**Pull request by josh146**
Updates the interface documentation to include:

* Mention of the new `KerasLayer` and `TorchLayer` classes

* The new variable handling

* Previous modifications made to `qml.grad` and `qml.jacobian`

* The new NumPy logo

### CODE REVIEW

File: doc/introduction/interfaces.rst [hunk 1]
```diff
@@ -76,6 +76,18 @@ See the links below for walkthroughs of each specific interface:
...
+
+
+:html:`<div class="summary-table">`
+
+.. autosummary::
+
+    pennylane.qnn.KerasLayer
+    pennylane.qnn.TorchLayer
+
+:html:`</div>`
```
Comment (mariaschuld) [inline doc/introduction/interfaces.rst:90]: Nice!

Comment (josh146) [reply doc/introduction/interfaces.rst:90]: Was trying to work out the best way to make them 'front and center' without duplicating the details Tom has already added to the docstrings

File: doc/introduction/circuits.rst [hunk 2]
```diff
@@ -117,7 +117,8 @@ devices, the options are:
```
Comment (mariaschuld) [inline doc/introduction/circuits.rst:117]: Can't comment, but maybe the header in line 101 above should get the same formatting?

**Comment by josh146** [comment]
@mariaschuld, quick question: Currently the `Circuits` quickstart only discusses circuit construction. To make gradients easier to find, should we rename the `Interfaces` page to `Interfaces and gradients`? Or `Interfaces and training`?

image

File: doc/introduction/circuits.rst (hunk 2 above)
Comment (josh146) [reply doc/introduction/circuits.rst:117]: Good call, have made the change :+1:

File: doc/introduction/interfaces/numpy.rst [hunk 3]
```diff
@@ -33,13 +33,13 @@ Via the QNode decorator
...
 in PennyLane. By default, all QNodes are constructed for the NumPy interface,
-but this can also be specified explicitly by passing the ``interface='numpy'`` keyword argument:
+but this can also be specified explicitly by passing the ``interface='autograd'`` keyword argument:
 
 .. code-block:: python
 
     dev = qml.device('default.qubit', wires=2)
 
-    @qml.qnode(dev, interface='numpy')
+    @qml.qnode(dev, interface='autograd')
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:42]: (docstrings have long favoured using `'autograd'` over `'numpy'`, so I've made the update here as well)

File: doc/introduction/interfaces/numpy.rst [hunk 4]
```diff
@@ -174,74 +226,56 @@ How does automatic differentiation work in the case where the QNode returns mult
...
+>>> j1 = qml.jacobian(circuit5)
+>>> j1(params)
+array([[ 0.        , -0.98006658],
+       [-0.98006658,  0.        ]])
 
 
 The output of :func:`~.jacobian` is a two-dimensional vector, with the first/second element being
 the partial derivative of the first/second expectation value with respect to the input parameter.
-The Jacobian function has the same signature as the gradient function, requiring the user to specify
-which argument should be differentiated.
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:203]: Removed, as this is no longer the case post-variable refactor

File: doc/introduction/interfaces/numpy.rst [hunk 5]
```diff
@@ -174,74 +226,56 @@ How does automatic differentiation work in the case where the QNode returns mult
...
-    Currently, :func:`pennylane.jacobian` supports only the case where
-    ``argnum`` is a single integer. For quantum functions with multiple arguments,
-    use the above method of combining arguments in an array to get the full Jacobian matrix.
 
 
 Advanced Autograd usage
 -----------------------
 
-The PennyLane NumPy interface leverages the Python library `autograd <https://github.com/HIPS/autograd>`_ to enable automatic differentiation of NumPy code, and extends it to provide gradients of quantum circuit functions encapsulated in QNodes. In order to make NumPy code differentiable, Autograd provides a wrapped version of NumPy (exposed in PennyLane as :code:`pennylane.numpy`).
+The PennyLane NumPy interface leverages the Python library `autograd
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:262]: This section used the term 'user', and it shouldn't, so I gave it a quick re-write

File: doc/introduction/interfaces/numpy.rst [hunk 6]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
 Evaluating this gradient function at specific parameter values:
 
 >>> dcircuit(phi, theta)
 (array([ -4.79425539e-01,   1.11022302e-16]), array(0.0))
 
 
+Non-differentiable arguments
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Note that **all positional arguments provided to the QNode are assumed to be differentiable
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:124]:
```suggestion
Differentiable and non-differentiable arguments
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

How does PennyLane know which arguments of a quantum function are differentiable and which ones are not? 

As a basic rule, **all positional arguments provided to the QNode are assumed to be differentiable
```

File: doc/introduction/interfaces/numpy.rst [hunk 7]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+by default**:
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:125]:
```suggestion
by default, while keyword arguments are not.**:
```

Comment (mariaschuld) [reply doc/introduction/interfaces/numpy.rst:125]: Is that even true?

File: doc/introduction/interfaces/numpy.rst [hunk 8]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+
+>>> from pennylane import numpy as np
+>>> np.array([0.1, 0.2])
+tensor([0.1, 0.2], requires_grad=True)
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:129]:
```suggestion

```
However, you may want to pass arguments as positional arguments to a QNode but _not_ have PennyLane
consider them when computing gradients. To accomplish this, it is important to know that importing numpy from PennyLane has an important side effect: creating of arrays with ``np.array()`` produces 
a _tensor_ which behaves like "normal" numpy arrays, but has a special flag indicating whether it is trainable or not. 

>>> from pennylane import numpy as np
>>> np.array([0.1, 0.2])
tensor([0.1, 0.2], requires_grad=True)
```

File: doc/introduction/interfaces/numpy.rst [hunk 9]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+
+If you would like to provide explicit non-differentiable arguments to the
+QNode or gradient function, make sure to use a NumPy array that specifies
+``requires_grad=False``:
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:133]:
```suggestion
If you would like to provide explicit non-differentiable arguments to the
QNode or gradient function, make sure to import numpy via ``from pennylane import numpy as np``
and create a NumPy array that specifies
``requires_grad=False``:
```

File: doc/introduction/interfaces/numpy.rst [hunk 10]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+
+>>> np.array([0.1, 0.2], requires_grad=False)
+tensor([0.1, 0.2], requires_grad=False)
+
+The ``requires_grad`` argument can be passed to any NumPy function provided by
+PennyLane.
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:139]:
```suggestion
.. note::

  The ``requires_grad`` argument can be passed to any NumPy function provided by PennyLane.
```

File: doc/introduction/interfaces/numpy.rst [hunk 11]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+.. code-block:: python
+
+    dev = qml.device('default.qubit', wires=5)
+
+    @qml.qnode(dev)
+    def circuit(weights, data, wires):
+        qml.templates.AmplitudeEmbedding(data, wires=wires, normalize=True)
+        qml.RX(weights[0], wires=wires[0])
+        qml.RY(weights[1], wires=wires[1])
+        qml.RZ(weights[2], wires=wires[2])
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:153]:
```suggestion
```

File: doc/introduction/interfaces/numpy.rst [hunk 12]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+        qml.CNOT(wires=[wires[0], wires[1]])
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:154]:
```suggestion
```

Just to shorten?

Comment (josh146) [reply doc/introduction/interfaces/numpy.rst:154]: Ah, the reason I added these extra gates was because I kept getting trivial Jacobians otherwise 😆

File: doc/introduction/interfaces/numpy.rst [hunk 13]
```diff
@@ -110,14 +110,66 @@ with respect to both QNode parameters ``phi`` and ``theta``:
...
+        return qml.expval(qml.PauliZ(wires[0]))
+
+We must specify that ``data`` and ``wires`` are NumPy arrays with ``requires_grad=False``:
+
+>>> weights = np.array([0.1, 0.2, 0.3])
+>>> data = np.random.random([2**3], requires_grad=False)
+>>> wires = np.array([2, 0, 1], requires_grad=False)
+>>> circuit(weights, data, wires)
+0.16935626052294817
+
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:165]: Shall we add something along the lines of 

If you do not want to specify the `requires_grad` flag, simply create a QNode that accepts the non-differentiable arguments as keyword arguments:

[Example]

This will also work if you create arrays after importing vanilla numpy with ``import numpy as np``.

Comment (mariaschuld) [reply doc/introduction/interfaces/numpy.rst:165]: Maybe fits better at the end...

File: doc/introduction/interfaces/numpy.rst [hunk 14]
```diff
@@ -174,74 +226,56 @@ How does automatic differentiation work in the case where the QNode returns mult
...
->>> params = np.array([np.pi / 3, 0.25, np.pi / 2])
->>> j2(params)
-array([[-8.66025404e-01, -5.55111512e-17,  0.00000000e+00],
-       [-4.71844785e-16, -1.38777878e-17, -5.00000000e-01]])
-
-.. warning::
-
-    Currently, :func:`pennylane.jacobian` supports only the case where
-    ``argnum`` is a single integer. For quantum functions with multiple arguments,
-    use the above method of combining arguments in an array to get the full Jacobian matrix.
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:232]: Nice!

File: doc/introduction/interfaces/numpy.rst (hunk 7 above)
Comment (josh146) [reply doc/introduction/interfaces/numpy.rst:125]: Keyword arguments can never be differentiable in the current system :) I would like to change that, so that the type of argument no longer affects differentiability at all (hence why I never mentioned keyword args here, I did that somewhat deliberately)

File: doc/introduction/interfaces/numpy.rst (hunk 9 above)
Comment (josh146) [reply doc/introduction/interfaces/numpy.rst:140]: Have rewritten this section to include your changes (and have added the import line to the examples to make it explicit!)

**Comment by mariaschuld** [comment]
> @mariaschuld, quick question: Currently the `Circuits` quickstart only discusses circuit construction. To make gradients easier to find, should we rename the `Interfaces` page to `Interfaces and gradients`? Or `Interfaces and training`?
> 
> image
> 

Good point!

I would simply go with "Training circuits", and then explain in the landing page that this depends on the interface. In some sense the word "interface" is totally lost on the user here, because most people don't know what that is, they want to know how to train a circuit, not how the parts of PennyLane work together...
